testing.rst

```diff
@@ -139,249 +170,313 @@ tests as far as PHPUnit is concerned, but they have a very specific workflow:
-you'll use to crawl your site::
+    .. code-block:: php
 
-    $crawler = $client->request('GET', '/post/hello-world');
+        // config/packages/test/swiftmailer.php
 
-The ``request()`` method (read
-:ref:`more about the request method <testing-request-method-sidebar>`)
-returns a :class:`Symfony\\Component\\DomCrawler\\Crawler` object which can
-be used to select elements in the response, click on links and submit forms.
+        // ...
+        $container->loadFromExtension('swiftmailer', [
+            'disable_delivery' => true,
+        ]);
+
+You can also use a different environment entirely, or override the default
+debug mode (``true``) by passing each as options to the ``createClient()``
+method::
+
+    $client = static::createClient([
+        'environment' => 'my_test_env',
+        'debug'       => false,
+    ]);
 
 .. tip::
 
-    The ``Crawler`` only works when the response is an XML or an HTML document.
-    To get the raw content response, call ``$client->getResponse()->getContent()``.
+    It is recommended to run your test with ``debug`` set to ``false`` on
+    your CI server, as it significantly improves test performance.
 
-The crawler integrates with the ``symfony/css-selector`` component to give you the
-power of CSS selectors to find content in a page. To install the CSS selector
-component, run:
+Customizing Environment Variables
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-.. code-block:: terminal
+If you need to customize some environment variables for your tests (e.g. the
+``DATABASE_URL`` used by Doctrine), you can do that by overriding anything you
+need in your ``.env.test`` file:
 
-    $ composer require --dev symfony/css-selector
+.. code-block:: text
 
-Now you can use CSS selectors with the crawler. To assert that the phrase
-"Hello World" is present in the page's main title, you can use this assertion::
+    # .env.test
+    DATABASE_URL="mysql://db_user:db_password@127.0.0.1:3306/db_name_test?serverVersion=5.7"
 
-    $this->assertSelectorTextContains('html h1.title', 'Hello World');
+    # use SQLITE
+    # DATABASE_URL="sqlite:///%kernel.project_dir%/var/app.db"
 
-This assertion checks if the first element matching the CSS selector contains
-the given text. This assert calls ``$crawler->filter('html h1.title')``
-internally, which allows you to use CSS selectors to filter any HTML element in
-the page and check for its existence, attributes, text, etc.
+This file is automatically read in the ``test`` environment: any keys here override
+the defaults in ``.env``.
 
-The ``assertSelectorTextContains`` method is not a native PHPUnit assertion and is
-available thanks to the ``WebTestCase`` class.
+Configuring a Database for Tests
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-.. versionadded:: 4.3
+Tests that interact with the database should use their own separate database to
+not mess with the databases used in the other :ref:`configuration environments <configuration-environments>`.
```

Comment:
"mess" =)

Reply:
Yeah, original text, should be reworded :wink: (and also talk about the bad DX when using Docker + Symfony CLI + databases)